src/cephadm/cephadmlib/templating.py

```diff
@@ -25,6 +27,36 @@ def __repr__(self) -> str:
         return repr(self.value)
 
 
+class _PackageLoader(jinja2.PackageLoader):
+    """Workaround for PackageLoader when using cephadm with relative paths.
+
+    It was found that running the cephadm zipapp from a local dir (like:
+    `./cephadm`) instead of an absolute path (like: `/usr/sbin/cephadm`) caused
+    the PackageLoader to fail to load the template.  After investigation it was
+    found to relate to how the PackageLoader tries to normalize paths and yet
+    the zipimporter type did not have a normalized path (/home/foo/./cephadm
+    and /home/foo/cephadm respectively).  When a full absolute path is passed
+    to zipimporter's get_data method it uses the (non normalized) .archive
+    property to strip the prefix from the argument. When the argument is a
+    normalized path - the prefix fails to match and is not stripped and then
+    the full path fails to match any value in the archive.
+
+    This shim subclass of jinja2.PackageLoader tweaks the loader's .archive
+    property to *be* a normalized path, so that the prefixes can match and be
+    stripped from path to be looked up w/in the zip file.
+    """
+
+    def __init__(
+        self,
+        package_name: str,
+        package_path: str = 'templates',
+        encoding: str = 'utf-8',
+    ) -> None:
+        super().__init__(package_name, package_path, encoding)
+        if isinstance(self._loader, zipimport.zipimporter):
+            self._loader.archive = os.path.normpath(self._loader.archive)
```

Comment:
I think this error is in reference to this line
```
2023-11-15T07:06:32.711 INFO:teuthology.orchestra.run.smithi097.stderr:Traceback (most recent call last):
2023-11-15T07:06:32.810 INFO:teuthology.orchestra.run.smithi097.stderr:  File "/usr/lib64/python3.6/runpy.py", line 193, in _run_module_as_main
2023-11-15T07:06:32.811 INFO:teuthology.orchestra.run.smithi097.stderr:    "__main__", mod_spec)
2023-11-15T07:06:32.811 INFO:teuthology.orchestra.run.smithi097.stderr:  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
2023-11-15T07:06:32.811 INFO:teuthology.orchestra.run.smithi097.stderr:    exec(code, run_globals)
2023-11-15T07:06:32.811 INFO:teuthology.orchestra.run.smithi097.stderr:  File "/tmp/tmpyk8frwf6.cephadm.build/__main__.py", line 7851, in <module>
2023-11-15T07:06:32.811 INFO:teuthology.orchestra.run.smithi097.stderr:  File "/tmp/tmpyk8frwf6.cephadm.build/__main__.py", line 7825, in main
2023-11-15T07:06:32.812 INFO:teuthology.orchestra.run.smithi097.stderr:  File "/tmp/tmpyk8frwf6.cephadm.build/cephadmlib/logging.py", line 217, in cephadm_init_logging
2023-11-15T07:06:32.812 INFO:teuthology.orchestra.run.smithi097.stderr:  File "/tmp/tmpyk8frwf6.cephadm.build/cephadmlib/logging.py", line 238, in write_cephadm_logrotate_config
2023-11-15T07:06:32.812 INFO:teuthology.orchestra.run.smithi097.stderr:  File "/tmp/tmpyk8frwf6.cephadm.build/cephadmlib/templating.py", line 90, in render
2023-11-15T07:06:32.812 INFO:teuthology.orchestra.run.smithi097.stderr:  File "/tmp/tmpyk8frwf6.cephadm.build/cephadmlib/templating.py", line 75, in _env
2023-11-15T07:06:32.812 INFO:teuthology.orchestra.run.smithi097.stderr:  File "/tmp/tmpyk8frwf6.cephadm.build/cephadmlib/templating.py", line 81, in _loader
2023-11-15T07:06:32.813 INFO:teuthology.orchestra.run.smithi097.stderr:  File "/tmp/tmpyk8frwf6.cephadm.build/cephadmlib/templating.py", line 59, in __init__
2023-11-15T07:06:32.813 INFO:teuthology.orchestra.run.smithi097.stderr:AttributeError: readonly attribute
```
https://pulpito.ceph.com/adking-2023-11-15_03:42:13-orch:cephadm-wip-adk-testing-2023-11-14-1702-distro-default-smithi/7458721
https://pulpito.ceph.com/adking-2023-11-15_03:42:13-orch:cephadm-wip-adk-testing-2023-11-14-1702-distro-default-smithi/7458695
https://pulpito.ceph.com/adking-2023-11-15_03:42:13-orch:cephadm-wip-adk-testing-2023-11-14-1702-distro-default-smithi/7458747

I believe all tests that failed with this (all that failed pulling an image from https://pulpito.ceph.com/adking-2023-11-15_03:42:13-orch:cephadm-wip-adk-testing-2023-11-14-1702-distro-default-smithi/) were either running on centos 8.stream, or rhel 8.6

Reply:
can confirm, on a centos 9.stream VM I had no issue, but as soon as I used this on cnetos 8.stream I got
```
[root@vm-00 ~]# cephadm list-networks
Traceback (most recent call last):
  File "/usr/lib64/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/sbin/cephadm/__main__.py", line 7851, in <module>
  File "/usr/sbin/cephadm/__main__.py", line 7825, in main
  File "/usr/sbin/cephadm/cephadmlib/logging.py", line 217, in cephadm_init_logging
  File "/usr/sbin/cephadm/cephadmlib/logging.py", line 238, in write_cephadm_logrotate_config
  File "/usr/sbin/cephadm/cephadmlib/templating.py", line 90, in render
  File "/usr/sbin/cephadm/cephadmlib/templating.py", line 75, in _env
  File "/usr/sbin/cephadm/cephadmlib/templating.py", line 81, in _loader
  File "/usr/sbin/cephadm/cephadmlib/templating.py", line 59, in __init__
AttributeError: readonly attribute

```

Reply:
Lovely. Once again centos8/python3.6 rears its ugly head. I need to remember to test this kind of hackery on the oldest and newest platforms. Worst case scenario: I can write a complete template loader rather than trying to "patch-around" issues with the existing loader, but it's frustrating.

Reply:
I have pushed a new version that doesn't try to change properties of existing types. I think this should work on a wider variety of python versions. It's a bit longer though, which is too bad.

Reply:
confirmed this works for both:
* SLES15SP3 / python3.6
* openSUSE tumbleweed / python3.11